membership-attribute-service/app/services/AttributesFromZuora.scala

```diff
@@ -58,10 +63,11 @@ class AttributesFromZuora(implicit val executionContext: ExecutionContext) exten
       val maybeAttributes = maybeZAttributes.map(ZuoraAttributes.asAttributes(_))
       attributesFromDynamo.foreach(maybeUpdateCache(maybeZAttributes, _, maybeAttributes))
       Future.successful("Zuora", maybeAttributes)
-    }).leftMap { zuoraError => // fallback to Dynamo if Zuora fails
-      SafeLogger.error(scrub"Failed to retrieve attributes from Zuora for $identityId because $zuoraError so falling back on cache.")
-      attributesFromDynamo.map(maybeDynamoAttributes => ("Dynamo", maybeDynamoAttributes.map(DynamoAttributes.asAttributes(_))))
-    }.merge.flatten // both branches return same type so can be merged
+    }) // fallback to cache if Zuora networking error or un-parsable response
+      .leftMap(fallbackToCache) // handles un-parsable response etc.
+      .merge // both branches return the same type so we can merge
+      .flatten
+      .recoverWith { case e => fallbackToCache(e.toString) } // handles timeouts etc.
```

Comment:
Worth putting a test to ensure we don't have this mistake again here?

Reply:
I'll give it a go. It might be fiddly.

Reply:
I've added the test. Also manually traced with debugger to make sure fallback executes. I can see now that without this fix exception was indeed thrown which would then be caught by Play's top level error handler which converts it to 500, however fallback would not be executed.